ddtrace/ddtrace.go

```diff
@@ -79,6 +79,17 @@ type SpanContext interface {
 	ForeachBaggageItem(handler func(k, v string) bool)
 }
 
+// ReadWriteSpan implementations are spans which can be read from and modified by using the provided methods.
+type ReadWriteSpan interface {
+	Span
+
+	// Tag returns the tag value held by the given key, nil if none was found.
+	Tag(key string) interface{}
+
+	// IsError reports wether the span is an error.
+	IsError() bool
+}
+
```

Comment:
We probably want to change this to a struct rather than an interface, and also have it live inside `ddtrace/tracer` rather than `ddtrace`.

Making this an interface means that we will never be able to add or remove methods without bumping the major version number. This is something that's caused problems for us with the `Span` and `Tracer` interfaces, so we should try and avoid it here. 

We don't really need an interface since we don't expect users to implement their own `ReadWriteSpan` and instead, this is a type we will pass to them. Making it a struct type means we will be able to grow the API as needed without breaking backwards compatibility.

The other thing here is that we likely don't want to embed the `Span` interface. Doing so means we inherit all of `Span`'s methods, which don't all make sense in the pipeline context. For instance, the presence of the `Finish()` method on `ReadWriteSpan` implies that `span.Finish()` is a valid call to make inside the pipeline, when in fact it is not. Although we could make sure `span.Finish()` is not harmful, it's better to not include it at all so as not to pollute the API we present to the users, and only include exactly the set of methods we want them to use.

Reply:
Thanks for the review and feedback 🙂  This approach makes sense to me, working on the changes!